nodestream/pipeline/pipeline.py

```diff
@@ -91,22 +194,39 @@ class Pipeline(AggregatedIntrospectiveIngestionComponent):
     def __init__(self, steps: List[Step], step_outbox_size: int) -> None:
         self.steps = steps
         self.step_outbox_size = step_outbox_size
+        self.logger = getLogger(self.__class__.__name__)
+        self.errors = []
 
     async def run(
         self, progress_reporter: Optional[PipelineProgressReporter] = None
     ) -> AsyncGenerator[Any, Any]:
         current_executor = None
         tasks = []
 
-        for step in self.steps:
+        for step_index, step in enumerate(self.steps):
             current_executor = StepExecutor(
-                upstream=current_executor, step=step, outbox_size=self.step_outbox_size
+                upstream=current_executor,
+                step=step,
+                outbox_size=self.step_outbox_size,
+                step_index=step_index,
             )
             tasks.append(asyncio.create_task(current_executor.work_loop()))
 
         current_executor.set_end_of_line(progress_reporter)
-
-        await asyncio.gather(*tasks)
+        return_states = await asyncio.gather(*tasks, return_exceptions=True)
+        for return_state in return_states:
+            if return_state:
+                self.errors.append(return_state)
+
+        # Raise the error and log it.
+        if self.errors:
+            try:
+                raise PipelineException(self.errors)
+            except PipelineException:
+                self.logger.exception(
+                    "Exception during execution of the pipeline.",
+                )
+                raise
```

Comment:
We already log the inner exceptions as they happen. Do we need this log message?